docs/enhancements/image-handling-service.md

```diff
@@ -0,0 +1,155 @@
+---
+title: image-handling-service
+authors:
+  - "@carbonin"
+creation-date: 2021-07-06
+last-updated: 2021-07-07
+---
+
+# Image Handling Service
+
+## Summary
+
+Creating and serving the discovery image is a resource intensive operation which is
+significantly different from the other responsibilities of the assisted service.
+This should be handled by a separate service rather than the main application.
+
+## Motivation
+
+Most the assisted service functionality is based on many quick API calls with
+relatively small responses. Handling images is all about larger chunks of data and
+potentially longer operations. Splitting these responsibilities would allow us
+to manage and scale these deployments separately.
+
+### Goals
+
+* Create and serve discovery ISO images from a separate service
+* Make minimal changes to existing (or proposed) APIs
+
+### Non-Goals
+
+* Serve a discovery image directly to a BMC
+
+## Proposal
+
+The new service (image service) will handle customizing and serving the customized
+image to users. It will query assisted service to retrieve the information that needs
```

Comment:
My original thought was to create the ignition on the fly in the image service.

What to do about the discovery ignition is on the list of open questions though. Having the service generate the ignition and save it would simplify the new service a bit as it would limit the information we need to get from assisted, but it's not just the ignition we need to embed. The minimal iso also adds an additional initrd image for static networking cases, I don't think we'd want to handle that in the same way.

I'll have to think about this a bit more and will update the proposal with something more concrete.

Reply:
The discovery ignition should still be available for download by users via
`GET /v1/clusters/{cluster_id}/downloads/files?file_name=discovery.ign`

The image service can call the same URL to retrieve the ignition

Reply:
Yeah, I'm thinking keeping the ignition definition in the assisted service would work better than moving all of that logic into the image service. This would also resolve the issue with exposing the pull secret over the API that @avishayt mentioned in https://github.com/openshift/assisted-service/pull/2172#discussion_r665060989.

We could probably add a download link for the initrd as well (not sure it would be worth saving this). That would make the image service very generic, which is also nice.

Reply:
Is there a use case that require ignition per cluster?

Reply:
That's our use case currently, right @filanov?
Not sure I understand the question.

Reply:
Asking because maybe it's not a real requirement, currently we are doing that because the image is under cluster, the question is if we have a case that require specific configuration to a single cluster. 
With late binding for example multiple clusters will use the same InfraEnv so it doesn't look like we have the option to add specific ignition to a cluster. 
We know that Infra env will support single architecture (arm/x86) so in this case only clusters from specific architectures will be able to use this ignition.

Reply:
Maybe it's not needed, just wanted to make sure that this is indeed the case

Reply:
I'm not sure it would change anything related to this proposal.

If we need different images per cluster then we'd need a different `InfraEnv` to store the details about that image, right?
Unless you're thinking about only allowing specific changes? For example [we've talked about](https://github.com/openshift/assisted-service/pull/2172#issuecomment-878726871) possibly adding the OCP version as a parameter to this service rather than putting it in the infra env, but decided that wouldn't be a good idea. Are you thinking we could do something like that for arch? I think the same logic might apply in that situation.